### PR TITLE
janet/APPEALS-10065 upload endpoint

### PR DESCRIPTION
**Pull request by j-n-t-l**
Resolves [APPEALS-10065](https://vajira.max.gov/browse/APPEALS-10065)

### Description
The Caseflow IDT API has an endpoint that can upload a .pdf to a Veteran's eFolder when there is no appeal
The existing IDT endpoint for uploading a document to VBMS is `/idt/api/v1/appeals/{{appealId}}/upload_document` but was changed to `/idt/api/v1/upload_document` so only the veteran file number is needed. This is because a veteran is not always tied to an appeal.

### Acceptance Criteria

- [ ] IDT user has the ability to upload a .pdf to Veteran's eFolder using the Veteran's file number 
- [x] IDT user has the ability to upload a .pdf to Veteran's eFolder using the Veteran's SSN
- [ ] IDT user has the ability upload a .pdf to Veteran's eFolder using the external id (BFKEY or UUID) of a past or present appeal (Current Functionality)
- [ ] A new IDT endpoint is created
    - URL Format: idt/api/v1/upload_document
- [ ] When the endpoint queues the upload successfully the endpoint returns 201
- [ ] When the endpoint receives a bad idt token return 403 Auth error as other IDT end point do
- When the endpoint fails to find the veteran the endpoint returns a 400 with a json payload:
    - “error_id” (UUID for the error that can be found in the logs by support)
    - “error_message” “The veteran was unable to be found.”
- [ ] When the endpoint fails to find the appeal the endpoint returns a 400 with a json payload:
    - “error_id” (UUID for the error that can be found in the logs by support)
    - “error_message” “The appeal was unable to be found."
- [ ] When the endpoint has a server error the endpoint returns a 500 with a json payload:
    - “error_id” (UUID for the error that can be found in the logs by support)
    - “error_message”  “There was an error on the server with your request. Please use the error_id to contact the help desk.”
- [ ] The endpoint will receive one of 3 look up ids
    - Appeal Id – The Vacols or AMA Appeal id if the veteran has an appeal in Caseflow
    - Veteran File Number – The veteran file number that must exists in VBMS
    - Veteran Social Security Number – The Veteran Social Security Number used to look up the Veteran file number
- [ ] The endpoint will look at the look up ids in the following order:
    - If appeals id is sent it WILL NOT Look at the other two ids
    - If appeals id is not present and veteran file number exists in the post request the endpoint will not look at the veteran SSN
    - If the appeal id nor the veteran file number is in the post request then the ssn will be used to find the veteran file number
- [ ] The IDT API end point URL accept a POST request with a JSON payload containing the following keys:
    - "veteran_file_number” – contains the Veteran File Number (as a string).
    - "appeal_id" - contains the external id (BFKEY or UUID) for the appeal.
    - “veteran_ssn” – contains the veteran social security number (as a string)
    - "document_type" - contains the VBMS document type
    - "document_name" - contains the name of the document for VBMS and log (Document Title in VBMS viewer)
    - "document subject" - contains the VBMS subject entry. May be omitted, in which case use the document_type as the document subject
    - "file" : base64 encoded pdf file to upload
- [ ] The end point must be able to accept files of at least 100mb before encoding
- [ ] Any valid VBMS document types can be uploaded into a Veterans eFolder. The following should be tested.
    - BVA Letter
    - BVA Decision
    - BVA-General
    - Hearing Related
    - Hearing Request
    - AMA Notification Letter
- [ ] The following information is maintained for auditing and reporting on ALL IDT eFolder uploads (where Veteran has no appeal/where Veteran has appeal):
    - When a document is uploaded to efolder
    - Document type
    - Document subject
    - Document name/title
    - Veteran information
    - File number
    - Who uploaded document
    - File size

### Testing Plan
Found at [APPEALS-10280](https://vajira.max.gov/browse/APPEALS-10280)
### Database Changes

* [ ] Added `veteran_file_number` as column to `vbms_uploaded_documents` table
* [ ] Added `veteran_file_number` as index to `vbms_uploaded_documents` table
* [ ] Removed `appeal_id` and `appeal_type` as column to `vbms_uploaded_documents` table
* [ ] Removed `appeal_id` and `appeal_type` as index to `vbms_uploaded_documents` table

### Integrations: Adding endpoints for external APIs
* [ ] Check that Caseflow's external API code for the endpoint matches the code in the relevant integration repo
  * [ ] Request: Service name, method name, input field names
  * [ ] Response: Check expected data structure
* [ ] Update Fakes
* [ ] Integrations impacting functionality are tested in Caseflow UAT

### CODE REVIEW

File: app/controllers/idt/api/v1/upload_vbms_document_controller.rb [hunk 1]
```diff
@@ -14,7 +14,53 @@ class Idt::Api::V1::UploadVbmsDocumentController < Idt::Api::V1::BaseController
     end
   end
 
+  def log_error(error)
+    Raven.capture_exception(error)
```
Comment (j-n-t-l) [inline app/controllers/idt/api/v1/upload_vbms_document_controller.rb:18]: remove raven call

File: app/controllers/idt/api/v1/upload_vbms_document_controller.rb [hunk 2]
```diff
@@ -14,7 +14,53 @@ class Idt::Api::V1::UploadVbmsDocumentController < Idt::Api::V1::BaseController
...
+    Rails.logger.error(error)
+  end
+
+  def bgs
+    @bgs ||= BGSService.new
+  end
+
   def create
+    # Find veteran from veteran file nummber of ssn
+    if request.parameters["veteran_file_number"].present?
```
Comment (j-n-t-l) [inline app/controllers/idt/api/v1/upload_vbms_document_controller.rb:28]: appeal id, file number, ssn order

File: app/controllers/idt/api/v1/base_controller.rb [hunk 3]
```diff
@@ -27,6 +27,20 @@ class Idt::Api::V1::BaseController < ActionController::Base
...
 
+  rescue_from Caseflow::Error::VeteranNotFound do |error|
+    message = " The veteran was unable to be found."
+    exception_uuid = SecureRandom.uuid
+    render(json: { message: "IDT Exception ID: " + exception_uuid + message }, status: :bad_request)
+    log_error(error)
+  end
+
+  rescue_from Caseflow::Error::AppealNotFound do |error|
+    message = " The appeal was unable to be found."
```
Comment (jshields13) [inline app/controllers/idt/api/v1/base_controller.rb:38]: don't need this just pass through the error message

File: app/jobs/upload_document_to_vbms_job.rb [hunk 4]
```diff
@@ -29,7 +29,8 @@ def add_context_to_sentry
     end
     Raven.extra_context(
       vbms_uploaded_document_id: document.id,
-      case_details_path: "/queue/appeals/#{document.appeal.external_id}"
+      # case_details_path: "/queue/appeals/#{document.appeal.external_id}"
```
Comment (jshields13) [inline app/jobs/upload_document_to_vbms_job.rb:32]: add the path in there

**Review by jshields13 — APPROVED** [review]

**Pull request merged**